billiford/go-clouddriver
# git/repo artifacts with token authorization

Adds support for the `git/repo` artifact type for credential types:
- public (no credentials)
- personal access token

Because retrieval of the code from github is via an HTTP call, SSH authorization is not supported.  SSH can be supported via `git clone`, but the [go-git library](https://github.com/go-git/go-git) has limitations, namely you need to specify the reference type (branch or tag).  To support cloning by commit hash, you need to also include the branch reference.

To use `go-git`, either 
1. The user needs to specify something different for the `branch` value in the UI for `go-clouddriver` than `clouddriver`, especially for a commit hash reference, such as
  • `refs/head/masterr`
  • `refs/tag/v1.0.0`
  • `d8c4d84e2dcc0c9d93aa89c10025fb9d9edffe41@git-repo`
2. The user still just specifies the branch or tag value for `go-clouddriver` as they would for `clouddriver`, and the code can first try to pull it as a branch, and if not found, then try it a tag - but commit hash would not be supported.  This has been implemented in the [git-repo-ssh branch](https://github.com/homedepot/go-clouddriver/blob/0d902a4f9e640c998e0ed8feeb9113e3ba1ad586/pkg/http/core/artifacts.go#L224)

Alternatively, `git` could be added to the `go-clouddriver` container and the code could call out to it to provide the same functionality and authorizations as `clouddriver`

While this PR does not provide the full functionality as `clouddriver`, it does provide all the functionality currently in use by the home depot.

To test this PR locally

1. Create artifact account config file `/opt/spinnaker/artifacts/config/git-repo.json`
```json
{
  "baseURL": "https://github.com",
  "name": "git-repo",
  "token": "<personal access token>", // optional
  "types": [
    "git/repo"
  ]
}
```
2. Start `go-clouddriver`
```
make run
```
3. Make request
```
curl "http://localhost:7002/artifacts/fetch" \
     -s -L -X PUT -H 'Content-Type: application/json' -d '{
	   "version": "master",
	   "artifactAccount": "git-repo",
	   "reference": "https://github.com/homedepot/go-clouddriver",
	   "type": "git/repo"
	 }' | tar t -
```

Should see results like
```
.github/
.github/workflows/
.github/workflows/go.yml
.gitignore
Makefile
README.md
api/
api/swagger.yaml
cmd/
cmd/clouddriver/
cmd/clouddriver/clouddriver.go
docker/
docker/Dockerfile
docker/build.sh
...
pkg/sql/sqlfakes/
pkg/sql/sqlfakes/fake_client.go
pkg/task.go
```

This [smoke test](https://sandbox-east.spinnaker.homedepot.com/#/applications/smoketests/executions?pipeline=gke-kustomize) shows this change working in the sandbox environment.
